### PR TITLE
[SPARK-7863][Core] Create SimpleDateFormat for every SimpleDateParam instance because it's not thread-safe

### PR DESCRIPTION
**Pull request by zsxwing**
SimpleDateFormat is not thread-safe. This PR creates new `SimpleDateFormat` for each `SimpleDateParam` instance.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/status/api/v1/SimpleDateParam.scala [hunk 1]
```diff
@@ -42,14 +42,16 @@ private[v1] class SimpleDateParam(val originalValue: String) {
 
 private[v1] object SimpleDateParam {
 
-  val formats: Seq[SimpleDateFormat] = {
+  // `SimpleDateFormat` is not thread-safe
+  val formats = new ThreadLocal[Seq[SimpleDateFormat]]() {
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/status/api/v1/SimpleDateParam.scala]: I wouldn't bother with `ThreadLocal`. Just make new objects on demand. These aren't expensive.

File: core/src/test/scala/org/apache/spark/status/api/v1/SimpleDateParamTest.scala [hunk 2]
```diff
@@ -22,7 +22,7 @@ class SimpleDateParamTest extends FunSuite with Matchers {
 
   test("date parsing") {
     new SimpleDateParam("2015-02-20T23:21:17.190GMT").timestamp should be (1424474477190L)
-    new SimpleDateParam("2015-02-20T17:21:17.190CST").timestamp should be (1424474477190L)
+    new SimpleDateParam("2015-02-20T17:21:17.190MDT").timestamp should be (1424474477190L)
```
Comment (srowen) [inline core/src/test/scala/org/apache/spark/status/api/v1/SimpleDateParamTest.scala]: Aren't you proposing a change to this test already?

Comment (zsxwing) [inline core/src/test/scala/org/apache/spark/status/api/v1/SimpleDateParamTest.scala]: Do you mean I should mention it in the jira?

File: core/src/main/scala/org/apache/spark/status/api/v1/SimpleDateParam.scala (hunk 1 above)
Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/status/api/v1/SimpleDateParam.scala]: I just feel using `ThreadLocal` is easy and minimizes the changes.

Comment (srowen) [inline core/src/main/scala/org/apache/spark/status/api/v1/SimpleDateParam.scala]: It also means you keep per-thread state lying around. I know, it's not big, but I don't think it's worth it.

File: core/src/test/scala/org/apache/spark/status/api/v1/SimpleDateParamTest.scala (hunk 2 above)
Comment (srowen) [inline core/src/test/scala/org/apache/spark/status/api/v1/SimpleDateParamTest.scala]: Ah sorry, it wasn't your PR; I mean this PR is making a similar but different change: https://github.com/apache/spark/pull/6377  We'll probably pull that in shortly anyway

**Comment by zsxwing** [comment]
@srowen updated as per your suggestion

File: core/src/main/scala/org/apache/spark/status/api/v1/SimpleDateParam.scala [hunk 3]
```diff
@@ -25,8 +25,19 @@ import javax.ws.rs.core.Response.Status
 import scala.util.Try
 
 private[v1] class SimpleDateParam(val originalValue: String) {
+
+  private val formats: Seq[SimpleDateFormat] = {
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/status/api/v1/SimpleDateParam.scala]: Now this is a field though. There's no need for that. The format objects can be created on demand to parse the input. If this `SimpleDateParam` object is short-lived I suppose it does not matter much but I would have just inlined all this.

Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/status/api/v1/SimpleDateParam.scala]: Inlined `formats`.

File: core/src/main/scala/org/apache/spark/status/api/v1/SimpleDateParam.scala [hunk 4]
```diff
@@ -25,8 +25,18 @@ import javax.ws.rs.core.Response.Status
 import scala.util.Try
 
 private[v1] class SimpleDateParam(val originalValue: String) {
+
   val timestamp: Long = {
-    SimpleDateParam.formats.collectFirst {
+    val formats: Seq[SimpleDateFormat] = {
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/status/api/v1/SimpleDateParam.scala]: OK... but this seems like a very complicated way of saying, "parse with one format, and if it fails, parse with the other. Why not just boil this down while we're here? you can even avoid allocating at least one of the two objects in most cases.

File: core/src/main/scala/org/apache/spark/status/api/v1/SimpleDateParam.scala [hunk 5]
```diff
@@ -39,17 +49,3 @@ private[v1] class SimpleDateParam(val originalValue: String) {
     )
   }
 }
-
-private[v1] object SimpleDateParam {
-
-  val formats: Seq[SimpleDateFormat] = {
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/status/api/v1/SimpleDateParam.scala]: why not just make this a `def makeFormats` or something? The old code is quite readable.

**Comment by zsxwing** [comment]
Rewrote SimpleDateParam to avoid parsing twice

**Comment by zsxwing** [comment]
retest this please

**Comment by srowen** [comment]
This LGTM.

**Pull request closed without merging**